server/src/main/java/io/druid/server/coordinator/DruidCoordinatorRuntimeParams.java

```diff
@@ -303,9 +304,30 @@ public Builder withSegmentReplicantLookup(SegmentReplicantLookup lookup)
       return this;
     }
 
+    public Builder withDatasources(Map<String, VersionedIntervalTimeline<String, DataSegment>> sources)
+    {
+      this.dataSources = sources;
+      return this;
+    }
+
     public Builder withDatasources(Collection<ImmutableDruidDataSource> dataSourcesCollection)
     {
-      dataSources.addAll(Collections.unmodifiableCollection(dataSourcesCollection));
+      dataSourcesCollection.stream().forEach(
+          dataSource -> {
+            VersionedIntervalTimeline<String, DataSegment> timeline = dataSources.computeIfAbsent(
+                dataSource.getName(),
+                k -> new VersionedIntervalTimeline<>(String.CASE_INSENSITIVE_ORDER)
+            );
+
+            dataSource.getSegments().forEach(
+                segment -> timeline.add(
+                    segment.getInterval(),
+                    segment.getVersion(),
+                    segment.getShardSpec().createChunk(segment)
+                )
+            );
+          }
+      );
```

Comment:
is this always done or only when compaction is enabled and this information is needed ?

Reply:
This is always done. Currently, all segments of the all dataSources are being gathered to emit some statistics in ```DruidCoordinatorLogger```. I only changed the type of ```dataSources``` in ```DruidCoordinatorRuntimeParams``` from ```Set<ImmutableDataSource>``` to ```Map<String, VersionedIntervalTimeline<String, DataSegment>>``` to use this information in ```DruidCoordinatorSegmentCompactor``` as well. 

Another alternative is to keep the current type of ```dataSources``` in ```DruidCoordinatorRuntimeParams``` and convert it to ```Map<String, VersionedIntervalTimeline<String, DataSegment>>``` in ```DruidCoordinatorSegmentCompactor``` when it's needed, but this may increase the memory usage. As an extreme example, there might be a lot of segments and the compaction might be enabled for all dataSources. In that case, this alternative approach will double the memory consumption. I expect users to enable the automatic compaction for all their realtime dataSources as well as some batch dataSources, so the memory burden might not be small.

Reply:
its alright given specially since it was there already.